### PR TITLE
Final Submit

### PR DESCRIPTION
**Pull request by cgallinaro**

### CODE REVIEW

File: Problem1/src/problem1/Problem1.java [hunk 1]
```diff
@@ -65,7 +60,7 @@ public void looping4(int n) {
 
     /**
      * looping4
-     * RUNTIME COMPLEXITY IS: O(?)
+     * RUNTIME COMPLEXITY IS: O(log n)
```
Comment (cgallinaro) [inline Problem1/src/problem1/Problem1.java]: This loop is also O(n^2)

File: Problem1/src/problem1/Problem1.java [hunk 2]
```diff
@@ -77,7 +72,7 @@ public void looping5(int n) {
 
     /**
      * looping6
-     * RUNTIME COMPLEXITY IS: O(?)
+     * RUNTIME COMPLEXITY IS: O(n^3)
```
Comment (cgallinaro) [inline Problem1/src/problem1/Problem1.java]: the inner loop will iterate n^3 times.  Thus, the complexity is O(n^5).

File: Problem1/src/problem1/Problem1.java [hunk 3]
```diff
@@ -125,7 +120,7 @@ public int recursive2(int n) {
 
     /**
      * recursive3
-     * RUNTIME COMPLEXITY IS: O(?)
+     * RUNTIME COMPLEXITY IS: O(n)
```
Comment (cgallinaro) [inline Problem1/src/problem1/Problem1.java]: On every recursive call, the value of n is being reduced by half.  Thus, the complexity is O(log n).

File: Problem1/src/problem1/Problem1.java [hunk 4]
```diff
@@ -140,7 +135,7 @@ public int recursive3(int n) {
 
     /**
      * recursive4
-     * RUNTIME COMPLEXITY IS: O(?)
+     * RUNTIME COMPLEXITY IS: O(n)
```
Comment (cgallinaro) [inline Problem1/src/problem1/Problem1.java]: For the non-base case, every recursive call results in 2 additional recursive calls (where n is only being reduced by 1).  Thus, the complexity is O(2^n).

File: Problem2/test/problem2/Problem2Test.java [hunk 5]
```diff
@@ -0,0 +1,92 @@
...
+    public void testFiveDivisibleByThree() {
+        assertNotEquals(divisible, divisibleByThree(743));        
+    }
+    
+    /**
+     * Test of divisibleByThree method, of class Problem2.
+     */
+    @Test
+    public void testSixDivisibleByThree() {
+        assertNotEquals(divisible, divisibleByThree(1000));
```
Comment (cgallinaro) [inline Problem2/test/problem2/Problem2Test.java]: It is good that you are covering values that are divisible and values that are not divisible.  When testing recursive methods, make sure you also include tests to cover your base cases:
```
assertEquals(divisible, divisibleByThree(3))
assertEquals(divisible, divisibleByThree(6))
assertEquals(divisible, divisibleByThree(9))
```

File: Problem3/test/problem3/PersonTest.java [hunk 6]
```diff
@@ -0,0 +1,250 @@
...
+        Person a = new Person("996");
+        
+        ArrayList<Person> sortedArray = getPersons();
+        
+        insertionSort(sortedArray);
+
+        boolean sorted = true;
+
+        for (int i = 0; i < sortedArray.size() - 1; i++) {
+            if (sortedArray.get(i).compareTo(sortedArray.get(i+1)) > 0) {
```
Comment (cgallinaro) [inline Problem3/test/problem3/PersonTest.java]: As a general rule, you should avoid using the method under test to check for correctness.  Since the insertionSort method uses compareTo when sorting, you should avoid using compareTo when validating.  To validate your insertionSort method, you should simply iterate over the array and verify the person objects are in the order you expect.

File: Problem3/test/problem3/PersonTest.java [hunk 7]
```diff
@@ -0,0 +1,250 @@
...
+        x.setName("y");
+        
+        assertEquals("y", x.getName());
+    }
+
+    /**
+     * Test of getAge method, of class Person.
+     */
+    @Test
+    public void testGetAge() {
```
Comment (cgallinaro) [inline Problem3/test/problem3/PersonTest.java]: Good job writing tests for the entire class.

File: Problem4/test/problem4/Problem4Test.java [hunk 8]
```diff
@@ -0,0 +1,120 @@
...
+     */
+    @Test
+    public void testMain() {
+    }
+
+    /**
+     * Test of flowerSort method, of class Problem4.
+     */
+    @Test
+    public void testFlowerSortOne() {
```
Comment (cgallinaro) [inline Problem4/test/problem4/Problem4Test.java]: When writing test cases for sorting algorithms, remember to include edge cases:
- Empty array
- Array with a single element

File: Problem4/src/problem4/Problem4.java [hunk 9]
```diff
@@ -1,21 +1,76 @@
...
+        
+        flowerSort(array);
+        
+        System.out.println(Arrays.toString(array));
+    }
+    
     /**
-     * @param args the command line arguments
+     * Sorting algorithm to sort array into ascending order
+     * RUNTIME COMPLEXITY IS: O(n log n)
```
Comment (cgallinaro) [inline Problem4/src/problem4/Problem4.java]: Complexity is actually O(n^2).  Also, you forgot part of the question:
Compare this algorithm to insertion sort
- describe an array which would perform better using your algorithm (over insertion sort)
- describe an array which would perform better using insertion sort (over your algorithm)

**Comment by cgallinaro** [comment]
Mark:

Problem 1: 3/5
Problem 2: 4.5/5
Problem 3: 5/5
Problem 4: 8/10

Total: 20.5/25
